### PR TITLE
feat(vpc): vpc module

### PR DESCRIPTION
**Pull request by Monitob**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      - apt-get update
+      - apt-get install nginx -y
+      - systemctl enable --now nginx
+      - hostnamectl hostname ${format("srv-%d.%s", count.index, scaleway_vpc_private_network.main.name)}
+      - echo "Hello i'm $(hostname)!" > /var/www/html/index.nginx-debian.html
+      - reboot # Make sure static DHCP reservation catch up
+    EOT
+  }
+
+}
```
Comment (remyleone) [inline main.tf:159]: terraform fmt everywhere

File: versions.tf [hunk 2]
```diff
@@ -0,0 +1,11 @@
+terraform {
+  required_providers {
+    scaleway = {
+      source = "scaleway/scaleway"
```
Comment (remyleone) [inline versions.tf:4]: Should a version be pinned here?

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  description = "A list of tags to add to all resources"
+  type        = list(string)
+  default     = []
+}
+
+################################################################################
+# Public Subnets
+################################################################################
+
+variable "public_subnets" {
```
Comment (baalooos) [inline variables.tf:74]: Variable created but not used. 
is it WIP?

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  description = "A list of public subnets inside the VPC"
+  type        = list(string)
+  default     = []
+}
+
+variable "public_subnet_ip_prefixes" {
```
Comment (baalooos) [inline variables.tf:80]: Variable created but not used. 
is it WIP?

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    var.vpc_tags,
+    var.vpc_private_network_tags
+  )
+}
+
+### DHCP Space of VPC
+resource "scaleway_vpc_public_gateway_dhcp" "main" {
+  subnet               = "192.168.1.0/24"
+  address              = "192.168.0.1"
+  pool_low             = "192.168.0.2"
```
Comment (baalooos) [inline main.tf:42]: this should be a variable

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  pool_high            = "192.168.0.50"
```
Comment (baalooos) [inline main.tf:43]: this should be a variable

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  enable_dynamic       = true
```
Comment (baalooos) [inline main.tf:44]: this should be a variable

File: main.tf [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  push_default_route   = true
```
Comment (baalooos) [inline main.tf:45]: this should be a variable

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  push_dns_server      = true
```
Comment (baalooos) [inline main.tf:46]: this should be a variable

File: main.tf [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  dns_servers_override = ["192.168.0.1"]
```
Comment (baalooos) [inline main.tf:47]: this should be a variable

File: main.tf [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  dns_local_name       = scaleway_vpc_private_network.main.name
+  depends_on           = [scaleway_vpc_private_network.main]
+}
+
+### VPC Gateway Network
+resource "scaleway_vpc_gateway_network" "main" {
+  gateway_id         = scaleway_vpc_public_gateway.main.id
+  private_network_id = scaleway_vpc_private_network.main.id
+  dhcp_id            = scaleway_vpc_public_gateway_dhcp.main.id
+  cleanup_dhcp       = true
```
Comment (baalooos) [inline main.tf:57]: Not sure if we need a variable here?

File: main.tf [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  enable_masquerade  = true
```
Comment (baalooos) [inline main.tf:58]: Not sure if we need a variable here?

File: main.tf [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  depends_on         = [scaleway_vpc_public_gateway.main, scaleway_vpc_private_network.main, scaleway_vpc_public_gateway_dhcp.main, scaleway_instance_server.srv]
+
+}
+
+resource "scaleway_vpc_public_gateway_dhcp_reservation" "app" {
+  count              = var.scale
+  gateway_network_id = scaleway_vpc_gateway_network.main.id
+  mac_address        = scaleway_instance_server.srv[count.index].private_network.0.mac_address
+  ip_address         = format("192.168.0.%d", (60 + count.index))
```
Comment (baalooos) [inline main.tf:67]: This should be a variable

File: loadbalancer.tf [hunk 14]
```diff
@@ -0,0 +1,47 @@
+################################################################################
```
Comment (baalooos) [inline loadbalancer.tf:1]: nice to have it in a separate file, 
maybe we should do a stand alone module for it and import it here

File: variables.tf (hunk 3 above)
Comment (Monitob) [reply variables.tf:74]: Yes on WIP

File: versions.tf (hunk 2 above)
Comment (Monitob) [reply versions.tf:4]: Yes. Fixed in last commit

File: variables.tf [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+  default     = []
+}
+
+variable "zone" {
+  description = "Zone in which resources should be created"
+  type        = string
+  default     = null
+}
+
+variable "create" {
```
[the remark this answers is not in the document]
Comment (baalooos) [reply variables.tf:71]: Sometimes in test env you want to disable some stuff.
for instance you don't to create a new vpc in your test account so you disable it. 
this is an edge case but it cost nothing to support it

File: instance.tf
Comment (remyleone) [inline instance.tf:1]: Why does this file exists?

File: README.md [hunk 16]
```diff
@@ -1 +1,91 @@
-# terraform-scaleway-vpc-module
\ No newline at end of file
+# terraform-scaleway-vpc-module
+
+This module can be used to deploy a VPC with various resources. Common deployment examples can be found in [examples/]().
```
Comment (remyleone) [inline README.md:3]: Broken link?

File: variables.tf [hunk 17]
```diff
@@ -0,0 +1,185 @@
+################################################################################
```
Comment (remyleone) [inline variables.tf:1]: Could those different variables be split into different files?

Comment (Monitob) [reply variables.tf:1]: It looks that the linter don't like this ``` variable "public_gateway_bastion_enabled" should be moved from public_gateway_variables.tf to variables.tf ```

File: README.md [hunk 18]
```diff
@@ -1 +1,90 @@
-# terraform-scaleway-vpc-module
\ No newline at end of file
+# terraform-scaleway-vpc-module
```
Comment (quantumsheep) [inline README.md:1]:
```suggestion
# Scaleway VPC Terraform module
```

File: README.md [hunk 19]
```diff
@@ -1 +1,90 @@
...
+
+This module can be used to deploy a VPC with various resources. Common deployment examples can be found in [examples/](./examples).
```
Comment (quantumsheep) [inline README.md:3]:
```suggestion
Terraform module that can be used to deploy VPC resources on Scaleway. Common deployment examples can be found in [examples/](./examples).
```

File: README.md [hunk 20]
```diff
@@ -1 +1,90 @@
...
+
+## Usage
+
+The example below builds a simple VPC with a PGW, a Load Balancer and some instances behind it.
```
Comment (quantumsheep) [inline README.md:7]:
```suggestion
The example below provision a basic VPC with a Public Gateway and a Load Balancer with some instances behind it.
```

**Review by quantumsheep — CHANGES_REQUESTED** [review]

File: examples/simple/README.md [hunk 21]
```diff
@@ -0,0 +1,54 @@
+# Simple VPC
```
Comment (quantumsheep) [inline examples/simple/README.md:1]: Rename this example to "Basic VPC"

File: examples/simple/README.md [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+Configuration in this directory creates set of VPC resources which may be sufficient for development environment.
```
Comment (quantumsheep) [inline examples/simple/README.md:3]:
```suggestion
Configuration in this directory creates a set of VPC resources which may be sufficient for development environment.
```

File: examples/simple/README.md [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+
+There is a public, private subnet and DHCP server created per availability zone.
+
+This configuration in only for demonstration purposes. Normally, you need to specify only names or your custom CIDR configuration.
```
Comment (quantumsheep) [inline examples/simple/README.md:7]: I don't understand the last sentence

**Review by quantumsheep — CHANGES_REQUESTED** [review]

File: examples/simple/main.tf [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+    "terraform-scaleway-vpc-module"
+  ]
+}
+
+################################################################################
+# VPC Module
+################################################################################
+
+module "vpc" {
+  source = "../../"
```
Comment (quantumsheep) [inline examples/simple/main.tf:20]:
```suggestion
  source = "scaleway/vpc"
```

**Review by quantumsheep — CHANGES_REQUESTED** [review]

File: variables.tf [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+  default     = []
+}
+
+variable "timeouts" {
+  description = "Define maximum timeout for creating, updating, and deleting VPC resources"
+  type        = map(string)
+  default     = {}
+}
+
+variable "azs" {
```
Comment (quantumsheep) [inline variables.tf:28]:
```suggestion
variable "zones" {
```

`azs` are not that easy to understand compared to `zones`

**Review by quantumsheep — CHANGES_REQUESTED** [review]

File: examples/simple/README.md (hunk 21 above)
Comment (Monitob) [reply examples/simple/README.md]: It's renamed. thanks for the feedback

File: examples/simple/README.md (hunk 23 above)
Comment (Monitob) [reply examples/simple/README.md]: the phrase has been reworked. Thanks

File: examples/simple/main.tf (hunk 24 above)
Comment (Monitob) [reply examples/simple/main.tf]: Will rollback this. I got an error `The module address "scaleway/vpc" could not be resolved.
`

**Review by quantumsheep — APPROVED** [review]
